### PR TITLE
feat: Add per-container redeploy button

### PR DESCRIPTION
**Pull request by mkaltner**
## Summary

Adds a **Redeploy** button for individual containers, positioned between **Restart** and **Remove** in the container actions menu. When clicked, the container is redeployed with the latest image while preserving its configuration (name, labels, networks, volumes, and restart policy).

Fixes: getarcaneapp/arcane#1553

---

### Backend (`backend/internal/services/container_service.go`)
- New `RedeployContainer` service method:
  - Inspects container to capture current configuration
  - Pulls latest image with timeout handling
  - Stops container if running (graceful 30s timeout)
  - Removes old container
  - Recreates container with cleaned networking config (strips ephemeral IP state)
  - Auto-starts if container was running or has restart policy
- Helper `buildCleanNetworkingConfigInternal`: Strips per-run network state (IPAddress, Gateway, etc.) to allow Docker to reassign IPs on recreation
- Logs container.deploy event for audit trail

### Backend (`backend/internal/huma/handlers/containers.go`)
- Registers new endpoint: `POST /environments/{id}/containers/{containerId}/redeploy`
- Handler follows standard container action pattern

### Frontend (`frontend/src/lib/services/container-service.ts`)
- Adds `redeployContainer` service method

### Frontend (`frontend/src/routes/(app)/containers/container-table.actions.ts`)
- Adds `handleRedeployContainer` action handler with confirmation dialog
- Adds `redeploy` entry to `containerActionConfigs` for use in table and detail views

### Frontend (`frontend/src/lib/components/action-buttons.svelte`)
- Adds container redeploy support to shared action buttons component
- Shows confirmation dialog before redeploying
- Redirects to containers list on success

### Frontend (`frontend/src/routes/(app)/containers/container-table.svelte`)
- Adds Redeploy button to container dropdown menu

### Frontend (`frontend/messages/en.json`)
- Adds container-specific confirmation and toast messages for redeploy action

---

## Behavior

- **Confirmation dialog** — Asks user to confirm before pulling and recreating
- **Name preservation** — Container recreated with same name (e.g. `myproject-redis-1`)
- **Network/volume preservation** — Same networks and volumes attached
- **Auto-start logic** — Starts if container was running OR has a restart policy (always/unless-stopped/on-failure)
- **Works for both** — Standalone containers and compose-managed containers

---

## PR Checklist

- [x] Code builds successfully in development environment (`./scripts/development/dev.sh start`)
- [x] Frontend loads at http://localhost:3000 with hot reload working
- [x] Backend responds at http://localhost:3552 with hot reload working
- [x] No linting errors
- [x] Formatting passes
- [x] Commit messages follow conventional format ✅
- [x] PR description explains the change and why it is needed ✅

## Manual Testing

- [x] Redeploy button appears in container actions menu
- [x] Confirmation dialog shows before redeploying
- [x] Container is recreated with same name after redeploy
- [x] Networks and volumes are preserved
- [x] Container auto-starts if it was running before redeploy
- [x] Works for both standalone and compose-managed containers
- [x] Redirects to containers list after successful redeploy

---

## AI Disclosure

**AI Tool:** Claude Code (Anthropic)  
**Assistance Level:** Significant — implementation was AI-generated with human direction, review, and testing oversight per [AI_POLICY.md](https://github.com/getarcaneapp/arcane/blob/main/AI_POLICY.md)

**Disclaimer** Greptiles Reviews use AI, make sure to check over its work.

To better help train Greptile on our codebase, if the comment is useful and valid **Like** the comment, if its not helpful or invalid **Dislike**

To have Greptile Re-Review the changes, mention `greptileai`.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds a per-container **Redeploy** action end-to-end: a new `POST /environments/{id}/containers/{containerId}/redeploy` endpoint backed by `ContainerService.RedeployContainer`, a CLI command, and frontend UI in both the container table and the detail view's action-buttons component. The backend logic is well-structured — it pulls the latest image, renames the old container to a timestamped backup name, stops it if running, recreates it with a cleaned networking config and cleared auto-generated hostname, honours restart-policy for auto-start, and attempts to restore the original container if any step fails. All prior review feedback (restart-policy check, hostname stripping, networking config sanitisation, event-log ordering, `io.Copy`/reader handling, toast message keys, navigation path, type safety) appears to have been addressed.

Key points:
- The overall flow is solid, with good rollback handling and comprehensive error logging at each step.
- `SanitizeContainerCreateEndpointSettingsForDockerAPI` only strips `MacAddress` for older Docker API versions; it performs a shallow copy and **does not strip `IPAMConfig`**. For containers created with an explicit static IP, this means the recreated container will request the same IP, which can race with Docker's IPAM release window. The new test exercises `Aliases` and the runtime-only `IPAddress` field but does not cover a static-IP (`IPAMConfig`) scenario — leaving this edge-case behaviour untested and undocumented.
- The frontend correctly uses `containerData.data.id` (not the previously flagged `data?.containerId`) to navigate to the new container's detail page after a successful redeploy.
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

- PR is safe to merge with one low-risk edge case to address in a follow-up.
- The implementation is well-structured with good error handling and rollback logic, and all prior critical review comments have been addressed. The only remaining concern is the lack of test coverage for the static-IP (`IPAMConfig`) scenario in `buildCleanNetworkingConfigInternal` — a rare edge case that could cause a `ContainerCreate` IP-conflict error for containers originally created with `--ip`. This is not a regression (the previous code had no redeploy support at all) but is worth addressing before the feature is considered fully hardened.
- backend/internal/services/container_service_test.go — needs a test case covering an endpoint with IPAMConfig set to document/verify static-IP handling.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| backend/internal/services/container_service.go | Core redeploy implementation: image pull with auth retry, rename-stop-create-start flow with rollback, restart policy honoured, hostname cleared, networking config sanitised. Overall well-structured; no new critical issues remaining after prior round of fixes. |
| backend/internal/services/container_service_test.go | New test only exercises the happy-path with Aliases + IPAddress but never sets IPAMConfig; since SanitizeContainerCreateEndpointSettingsForDockerAPI does a shallow copy, the behaviour for static-IP containers is untested and the IPAMConfig preservation/stripping contract is left implicit. |
| backend/internal/huma/handlers/containers.go | New RedeployContainer handler follows existing patterns correctly; returns full container details on success, minimal ID-only fallback if post-redeploy inspect fails. No issues found. |
| frontend/src/lib/components/action-buttons.svelte | Redeploy wired for both container and project types; post-success navigation uses containerData.data.id (correct path), separate toast messages per type, onActionComplete preserved for non-container paths. No remaining issues after prior fixes. |
| frontend/src/routes/(app)/containers/container-table.actions.ts | handleRedeployContainer added with confirm dialog, container-specific i18n keys, and refreshContainers() on success. containerActionConfigs.redeploy entry uses correct message keys. No issues found. |

</details>

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant UI as Frontend (Table / Detail)
    participant SVC as container-service.ts
    participant API as POST /containers/:id/redeploy
    participant CS as ContainerService.RedeployContainer
    participant D as Docker Daemon

    UI->>UI: confirmAction('redeploy') → confirm dialog
    UI->>SVC: redeployContainer(containerId)
    SVC->>API: POST /environments/{envId}/containers/{id}/redeploy
    API->>CS: RedeployContainer(ctx, containerID, user)

    CS->>D: ContainerInspect
    D-->>CS: containerInfo (config, state, networks)

    CS->>D: ImagePull(imageName) [with auth retry]
    D-->>CS: pull stream

    CS->>D: ContainerRename(id → backupName)
    CS->>D: ContainerStop(id) [if wasRunning, 30s timeout]

    CS->>D: ContainerCreate(name=original, config, hostConfig, cleanNetworks)
    D-->>CS: newContainerID

    alt shouldStart (wasRunning || restart policy)
        CS->>D: ContainerStart(newContainerID)
        alt start fails
            CS->>D: ContainerRemove(newContainerID, force)
            CS->>D: ContainerStart(originalID) [restore]
            CS->>D: ContainerRename(originalID → originalName) [restore]
            CS-->>API: error
        end
    end

    CS->>D: ContainerRemove(originalID, force) [cleanup old]
    CS->>CS: LogContainerEvent(container.deploy)
    CS-->>API: newContainerID

    API->>D: ContainerInspect(newContainerID)
    API-->>SVC: ContainerDetailsResponse
    SVC-->>UI: success
    UI->>UI: toast.success + goto(/containers/:newId)
```
</details>

<details open><summary><h3>Comments Outside Diff (7)</h3></summary>

1. `frontend/src/lib/components/action-buttons.svelte`, line 1258-1279 ([link](https://github.com/getarcaneapp/arcane/blob/128c13fc48f55b491cef0dafc517bf0be1195731/frontend/src/lib/components/action-buttons.svelte#L1258-L1279)) 

   **Project redeploy removed from dropdown — regression**

   Before this PR, the `{:else}` branch (which covers `type === 'project'` and other non-container types) contained a `<DropdownMenu.Item>` for redeploy. That item has been removed in this PR but nothing replaces it in the `else` branch, so any view that renders this component with a project/stack type will silently lose its **Redeploy** dropdown option.

   The inline-button paths (`ArcaneButton`) still correctly keep the `{:else}` redeploy button (lines ~893–899), so only the dropdown variant is broken.

   The fix depends on intent:
   - If the removal was meant to avoid the previous `goto('/containers')` navigation issue for projects, the underlying `onSuccess` handler in `confirmAction('redeploy')` should properly call `onActionComplete('running')` for non-container types (as the original code did), and the item should be reinstated.
   - If project redeploy is intentionally out of scope for this PR, the item removal should be explicitly reverted to avoid a silent regression.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: frontend/src/lib/components/action-buttons.svelte
   Line: 1258-1279

   Comment:
   **Project redeploy removed from dropdown — regression**

   Before this PR, the `{:else}` branch (which covers `type === 'project'` and other non-container types) contained a `<DropdownMenu.Item>` for redeploy. That item has been removed in this PR but nothing replaces it in the `else` branch, so any view that renders this component with a project/stack type will silently lose its **Redeploy** dropdown option.

   The inline-button paths (`ArcaneButton`) still correctly keep the `{:else}` redeploy button (lines ~893–899), so only the dropdown variant is broken.

   The fix depends on intent:
   - If the removal was meant to avoid the previous `goto('/containers')` navigation issue for projects, the underlying `onSuccess` handler in `confirmAction('redeploy')` should properly call `onActionComplete('running')` for non-container types (as the original code did), and the item should be reinstated.
   - If project redeploy is intentionally out of scope for this PR, the item removal should be explicitly reverted to avoid a silent regression.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. `frontend/src/lib/components/action-buttons.svelte`, line 1258-1279 ([link](https://github.com/getarcaneapp/arcane/blob/340eb12e1ceb9b297fc9a9371e846135de4b7a5d/frontend/src/lib/components/action-buttons.svelte#L1258-L1279)) 

   **Project/stack Redeploy item removed from second dropdown**

   The Redeploy `DropdownMenu.Item` was moved from the `{:else}` block (which covers `type === 'project'`, `type === 'stack'`, and other non-container types) into the `{#if type === 'container'}` block. As a result, non-container types have silently lost their Redeploy option in this second dropdown layout. The first dropdown (around line 1028) still correctly includes redeploy in its `{:else}` branch, so this regression is layout-specific.

3. `backend/internal/services/container_service.go`, line 266-271 ([link](https://github.com/getarcaneapp/arcane/blob/772f503a05e17f22a73c0ee7b0212fba082dae5c/backend/internal/services/container_service.go#L266-L271)) 

   **Bypasses Docker API multi-network compatibility layer**

   `dockerClient.ContainerCreate` is called directly here, but the codebase has a dedicated `libarcane.ContainerCreateWithCompatibility` helper that exists precisely for this scenario. The comment in `docker_compat.go` explains the issue:

   > On Docker API < 1.44 (e.g. Docker 24.x / API 1.43), `NetworkingConfig.EndpointsConfig` cannot contain more than one entry on `ContainerCreate`. `ContainerCreateWithCompatibility` handles this by keeping only the primary network in the initial create call and attaching the remaining networks via `NetworkConnect` before start.

   Compose-managed containers almost always have multiple user-defined networks. On a Docker 24.x host, redeploying such a container will fail with a daemon-side networking error — the old container will already have been removed with no recovery path.

   The upgrade path (`cli/upgrade/upgrade.go:425`) and the `inspectCompatibilityClient` wrapper both correctly use `ContainerCreateWithCompatibility`. This path should too:

   ```go
   createResp, err := libarcane.ContainerCreateWithCompatibility(ctx, dockerClient, client.ContainerCreateOptions{
       Config:           &newConfig,
       HostConfig:       containerInfo.HostConfig,
       NetworkingConfig: networkingConfig,
       Name:             containerName,
   })
   ```

   **Rule Used:** # Golang Pro
   
   Senior Go developer with deep expert... ([source](https://app.greptile.com/review/custom-context?memory=214b40a8-9695-4738-986d-5949b5d65ff1))

4. `backend/internal/services/container_service.go`, line 57-82 ([link](https://github.com/getarcaneapp/arcane/blob/51065ad3f953e10f6b1552cc75f840e7df0c79b1/backend/internal/services/container_service.go#L57-L82)) 

   **Static IP config silently lost on redeploy**

   `buildCleanNetworkingConfigInternal` copies only `Aliases` from each endpoint, discarding `IPAMConfig` entirely. This is correct for ephemeral/dynamic IP addresses (which was the motivation), but it also silently erases *intentional* static IPs a user may have configured with `--ip` or via a Compose `ipv4_address` field. After redeploy, those containers will receive a new dynamic IP, silently breaking services that rely on stable addresses (e.g. fixed DNS mappings, firewall rules, or peer services that hard-code the IP). The PR description states "preserves its configuration (name, labels, networks, volumes, and restart policy)" but network endpoint IP configuration is not preserved.

   Consider preserving non-empty `IPAMConfig` entries so explicit static IPs survive the redeploy:

   ```go
   endpointsConfig[netName] = &network.EndpointSettings{
       Aliases: append([]string(nil), ep.Aliases...),
       IPAMConfig: ep.IPAMConfig, // preserve intentional static IPs
   }
   ```

   If you want to keep the IPAMConfig only when it was explicitly set (i.e. non-empty addresses), you can conditionally include it:
   ```go
   var ipamCfg *network.EndpointIPAMConfig
   if ep.IPAMConfig != nil && (ep.IPAMConfig.IPv4Address != "" || ep.IPAMConfig.IPv6Address != "") {
       ipamCfg = ep.IPAMConfig
   }
   endpointsConfig[netName] = &network.EndpointSettings{
       Aliases:    append([]string(nil), ep.Aliases...),
       IPAMConfig: ipamCfg,
   }
   ```

5. `frontend/src/routes/(app)/containers/container-table.actions.ts`, line 65-70 ([link](https://github.com/getarcaneapp/arcane/blob/51065ad3f953e10f6b1552cc75f840e7df0c79b1/frontend/src/routes/(app)/containers/container-table.actions.ts#L65-L70)) 

   **`containerActionConfigs.redeploy` entry is dead code**

   `performContainerAction` is only ever called with `'start'`, `'stop'`, and `'restart'` (confirmed by `grep`). The `redeploy` case is instead handled by the dedicated `handleRedeployContainer` function which calls `containerService.redeployContainer` directly without going through this config. As a result, the `run`, `success`, and `failure` fields defined here are never executed.

   If the intent is purely documentation/centralization, a comment would be clearer. Otherwise, removing the entry would reduce confusion about the actual code path taken during a redeploy.

6. `backend/internal/services/container_service.go`, line 71-73 ([link](https://github.com/getarcaneapp/arcane/blob/edf1cac36889452fbeb979ac34bb774608975dfe/backend/internal/services/container_service.go#L71-L73)) 

   **Static `IPAMConfig` silently lost on redeploy**

   `buildCleanNetworkingConfigInternal` only copies `Aliases` into the new `EndpointSettings`, discarding `IPAMConfig` entirely. When a user has assigned a static IP to the container (e.g. `docker run --ip 192.168.100.2 --network my-net ...`), that preference is stored in `ep.IPAMConfig`. After a redeploy, Docker's IPAM will assign a random IP instead of the intended static one — silently breaking any dependency on a fixed address.

   The intent of the helper was to strip the *runtime-assigned* ephemeral fields (`IPAddress`, `Gateway`, etc.) while retaining user-declared configuration. `IPAMConfig` is user-declared and should be preserved:

   ```go
   endpointsConfig[netName] = &network.EndpointSettings{
       Aliases:    append([]string(nil), ep.Aliases...),
       IPAMConfig: ep.IPAMConfig, // preserve user-configured static IPs
   }
   ```

   **Rule Used:** # Golang Pro
   
   Senior Go developer with deep expert... ([source](https://app.greptile.com/review/custom-context?memory=214b40a8-9695-4738-986d-5949b5d65ff1))

7. `backend/internal/services/container_service_test.go`, line 77-94 ([link](https://github.com/getarcaneapp/arcane/blob/81f132fa8fcf6e644d257400441856fd04ea0bda/backend/internal/services/container_service_test.go#L77-L94)) 

   **Test doesn't cover `IPAMConfig` — static-IP redeploy behaviour left implicit**

   `SanitizeContainerCreateEndpointSettingsForDockerAPI` (called by `buildCleanNetworkingConfigInternal`) performs a **shallow copy** of each `EndpointSettings` struct — it only strips `MacAddress` for old API versions and otherwise preserves the endpoint as-is, including `IPAMConfig`.

   For containers originally created with an explicit static IP (`--ip 172.17.0.5`), Docker stores `IPAMConfig.IPv4Address = "172.17.0.5"` in the inspect response. Because the sanitizer preserves `IPAMConfig`, the recreated container would request the same static IP. Docker may reject this with an IP-conflict error during the brief window between the old container's removal and Docker reclaiming the address from its IPAM pool.

   The current test only exercises `Aliases` and the runtime-only `IPAddress` field (which Docker ignores during `ContainerCreate`), so this edge case is completely untested. Adding a sub-case with `IPAMConfig` set would both document the current behaviour explicitly and catch any future regression if the sanitizer is changed:

   ```go
   func TestBuildCleanNetworkingConfigInternalStripsIPAMConfig(t *testing.T) {
       staticIP := "172.18.0.5"
       containerInspect := container.InspectResponse{
           NetworkSettings: &container.NetworkSettings{
               Networks: map[string]*network.EndpointSettings{
                   "mynet": {
                       Aliases: []string{"svc"},
                       IPAMConfig: &network.EndpointIPAMConfig{
                           IPv4Address: staticIP,
                       },
                   },
               },
           },
       }

       out := buildCleanNetworkingConfigInternal(containerInspect, "1.44")
       require.NotNil(t, out)
       require.Contains(t, out.EndpointsConfig, "mynet")
       // IPAMConfig should be cleared so Docker reassigns the IP on recreation
       require.Nil(t, out.EndpointsConfig["mynet"].IPAMConfig)
   }
   ```

   If clearing `IPAMConfig` is the desired behaviour but is not yet implemented in `SanitizeContainerCreateEndpointSettingsForDockerAPI`, this test will fail and serve as a red flag. If preserving it is intentional (i.e. you want to re-request the same static IP and accept the race risk), the test should instead assert `IPAMConfig` is non-nil and document that decision with a comment.

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: backend/internal/services/container_service_test.go
Line: 77-94

Comment:
**Test doesn't cover `IPAMConfig` — static-IP redeploy behaviour left implicit**

`SanitizeContainerCreateEndpointSettingsForDockerAPI` (called by `buildCleanNetworkingConfigInternal`) performs a **shallow copy** of each `EndpointSettings` struct — it only strips `MacAddress` for old API versions and otherwise preserves the endpoint as-is, including `IPAMConfig`.

For containers originally created with an explicit static IP (`--ip 172.17.0.5`), Docker stores `IPAMConfig.IPv4Address = "172.17.0.5"` in the inspect response. Because the sanitizer preserves `IPAMConfig`, the recreated container would request the same static IP. Docker may reject this with an IP-conflict error during the brief window between the old container's removal and Docker reclaiming the address from its IPAM pool.

The current test only exercises `Aliases` and the runtime-only `IPAddress` field (which Docker ignores during `ContainerCreate`), so this edge case is completely untested. Adding a sub-case with `IPAMConfig` set would both document the current behaviour explicitly and catch any future regression if the sanitizer is changed:

```go
func TestBuildCleanNetworkingConfigInternalStripsIPAMConfig(t *testing.T) {
    staticIP := "172.18.0.5"
    containerInspect := container.InspectResponse{
        NetworkSettings: &container.NetworkSettings{
            Networks: map[string]*network.EndpointSettings{
                "mynet": {
                    Aliases: []string{"svc"},
                    IPAMConfig: &network.EndpointIPAMConfig{
                        IPv4Address: staticIP,
                    },
                },
            },
        },
    }

    out := buildCleanNetworkingConfigInternal(containerInspect, "1.44")
    require.NotNil(t, out)
    require.Contains(t, out.EndpointsConfig, "mynet")
    // IPAMConfig should be cleared so Docker reassigns the IP on recreation
    require.Nil(t, out.EndpointsConfig["mynet"].IPAMConfig)
}
```

If clearing `IPAMConfig` is the desired behaviour but is not yet implemented in `SanitizeContainerCreateEndpointSettingsForDockerAPI`, this test will fail and serve as a red flag. If preserving it is intentional (i.e. you want to re-request the same static IP and accept the race risk), the test should instead assert `IPAMConfig` is non-nil and document that decision with a comment.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: 81f132f</sub>

### CODE REVIEW

**Comment by mkaltner** [comment]
Reviewing and fixing comments.

File: backend/internal/services/container_service.go [hunk 1]
```diff
@@ -118,6 +118,91 @@ func (s *ContainerService) RestartContainer(ctx context.Context, containerID str
...
+	createResp, createErr := dockerClient.ContainerCreate(
+		ctx,
+		containerInspect.Config,
+		containerInspect.HostConfig,
+		&network.NetworkingConfig{
+			EndpointsConfig: containerInspect.NetworkSettings.Networks,
+		},
+		nil,
+		containerName,
+	)
```
[the remark this answers is not in the document]
Comment (mkaltner) [reply backend/internal/services/container_service.go]: Fixed

**Comment by mkaltner** [comment]
Done

**Comment by mkaltner** [comment]
I don't work with go/svelte much so sorry about the build errors.  I'll see if I can tackle them unless you can.  Thanks!

**Comment by kmendell** [comment]
No worries, Im just trying to get the bugs fixed right now, i just keep running the workflows for you in case, if theres something that you are not sure about just let me know.

**Comment by mkaltner** [comment]
Lets see if this push fixes it.  I shouldn't have trusted the AI code review, that made things worse.  lol

**Comment by kmendell** [comment]
Oh the moby errors are a PIA, im not sure why those would be happenig i havent loked fully though. and no worries :) you can run most of this locally so you dont rely on github workflows, most of the commands are in the Justfile but all can be ran without it too if you dont wnat to install [just](https://github.com/casey/just)

**Comment by mkaltner** [comment]
Yeah, that was my bad.  It looked good and I pushed it.  Getting VS Code setup for it now.  I'm a c#/azure dev by trade with little to no UI work.  :D

**Comment by kmendell** [comment]
Oh i dont know weather to apologize or praise you for dealing with that ;) just kidding. Take ur time, and ask if you have questions.

**Comment by kmendell** [comment]
You have to add the m.Common_redeploying() to the en.json file for i18n to clear those errors and what ever strings its complaining about.

File: frontend/src/lib/components/action-buttons.svelte [hunk 2]
```diff
@@ -407,18 +407,25 @@
...
 							result,
 							message: m.common_action_failed_with_type({ action: m.common_redeploy(), type }),
-							onSuccess: async () => {
+							onSuccess: async (data) => {
 								toast.success(m.common_redeploy_success({ type: name || type }));
-								onActionComplete('running');
+
+								const newId = data?.data?.newId;
+
+								if (newId) {
```
Comment (kmendell) [inline frontend/src/lib/components/action-buttons.svelte]: Wouldnt it be easier to just always redirect to the containerId no matter what? then it re-freshes the page vs checking if there's a new ID or not, The backend should just always return the containerId i think.

**Comment by mkaltner** [comment]
@greptileai Please re-review — all three issues from the previous review were addressed in commit 8e52ca20.

File: frontend/src/lib/components/action-buttons.svelte (hunk 2 above)
Comment (mkaltner) [reply frontend/src/lib/components/action-buttons.svelte]: Good call! ✅ Fixed in ac62b9ea:

- Backend now returns `containerId` in the response (`ContainerRedeployResponse` with `ContainerRedeployData`)
- Frontend redirects to `/containers/${data.containerId}` to stay on the detail page after redeploy
- Falls back to `/containers` list only if no ID is returned

Much better UX than kicking users back to the list. Thanks for the suggestion!

**Comment by mkaltner** [comment]
@greptileai Please re-review — all three original issues addressed in 8e52ca20, plus kmendell's feedback (return containerId and stay on detail page) implemented in ac62b9ea.

**Comment by mkaltner** [comment]
@greptileai All three issues from the second review addressed in eca88ce4:
1. Project redeploy navigation fixed (non-containers call onActionComplete)
2. Restart policy now honored (stopped containers with restart policies will start)
3. Timeout pointer now inline per codebase convention

Please re-review.

**Comment by mkaltner** [comment]
✅ **Re: Greptile feedback on cleanup logging**

Fixed in 67b89abf — cleanup failures when `ContainerStart` fails are now logged via `LogErrorEvent` (step: `cleanup_failed_start`) so operators have visibility into stopped-but-not-deleted containers.

Good catch on observability!

**Comment by mkaltner** [comment]
**Re: i18n keys**

✅ Fixed in cbcfd855 — changed the code to use `m.common_action_redeploying()` (which already exists in en.json) instead of adding a new `common_redeploying` key. Same pattern for `updating` and `removing` actions.

Thanks for catching that!

**Comment by mkaltner** [comment]
@greptileai Final re-review request — cleanup logging addressed in 67b89abf (your last identified style issue).

**Comment by mkaltner** [comment]
✅ **Addressed Greptile feedback in f3f379a6:**

**1. Deploy event logging (container_service.go:186-189)** — Moved to after `ContainerStart` succeeds. Now uses new container ID (`createResp.ID`) and treats logging errors as non-fatal (just warns). Prevents misleading audit trail and blocks from transient event-service issues.

**2. Stale hostname (container_service.go:253-258)** — Strips auto-generated hostname before recreating container. Creates shallow copy of config, clears `Hostname` if it matches old container's short ID. New container gets correct identity.

**3. TypeScript type safety (container-service.ts:86-89)** — Added `ContainerRedeployResponse` interface and changed return type from `Promise<any>`. Now properly typed with `containerId` validation.

All three correctness issues resolved.

**Comment by mkaltner** [comment]
@greptileai All three issues from the latest review addressed in f3f379a6:
1. Deploy event logging moved to success path (non-fatal)
2. Auto-generated hostname stripped before container recreation
3. TypeScript type safety added for redeploy response

Please re-review.

**Comment by mkaltner** [comment]
✅ **Fixed navigation bug in 73c8343b**

Changed `data?.containerId` → `data?.data?.containerId` in action-buttons.svelte to match the actual `ContainerRedeployResponse` structure. Navigation to redeployed container detail page now works correctly (was always falling back to `/containers` list).

**Comment by mkaltner** [comment]
@greptileai Navigation path fixed in 73c8343b — `data?.data?.containerId` now correctly accesses nested response structure. Please re-review.

**Comment by mkaltner** [comment]
I'm starting to have a love/hate relationship with greptile...  lol

**Comment by mkaltner** [comment]
✅ **Fixed TypeScript union type error in 9221d834**

- Cast `mutationFn` return to `Promise<ContainerRedeployResponse | Project>`
- Type-narrow `data` as `ContainerRedeployResponse` in `onSuccess` handler before accessing container-specific properties
- Verified locally: `svelte-check` passes with 0 errors

@greptileai Please re-review — Svelte Type Check CI failure resolved.

**Comment by mkaltner** [comment]
✅ **Fixed timeout pointer pattern in 128c13fc**

Replaced `func() *int { v := 30; return &v }()` with idiomatic temp variable:
```go
timeout := 30
_, err = dockerClient.ContainerStop(ctx, containerID, client.ContainerStopOptions{Timeout: &timeout})
```

@greptileai Please re-review.

**Comment by mkaltner** [comment]
✅ **Fixed i18n key inconsistencies in 5bcf6402**

- Changed confirm button label from `compose_pull_redeploy` → `common_redeploy`
- Changed toast in action-buttons to use `container_redeploy_success()` for container type
- Now consistent with container-table.actions.ts messaging

@greptileai Please re-review.

**Comment by mkaltner** [comment]
✅ **Fixed label mismatch** in commit 340eb12e

Changed line 653 in `container-table.svelte` from:
```svelte
{m.compose_pull_redeploy()}  // "Pull & Redeploy" (compose-project scope)
```

To:
```svelte
{m.common_redeploy()}  // "Redeploy" (single-container scope)
```

Now consistent with the confirm dialog label fixed in a prior commit.

@greptileai please re-review — should be ready to merge now

**Comment by mkaltner** [comment]
✅ **Project/stack redeploy regression fixed** in commit 772f503a

**Issue:** The `{:else}` branch (projects/stacks) in the second dropdown layout was missing the Redeploy menu item — silently breaking existing functionality.

**Fix:** Added the redeploy dropdown item back, matching the pattern used in:
- The other dropdown section (lines 1028-1030)
- The inline button section (lines 893-899)

Projects and stacks now have their Redeploy option restored in all dropdown variants.

@greptileai please re-review

**Comment by mkaltner** [comment]
✅ **Fixed Docker API compatibility** in commit 51065ad3

Changed line 266 from:
```go
createResp, err := dockerClient.ContainerCreate(ctx, ...)
```

To:
```go
createResp, err := libarcane.ContainerCreateWithCompatibility(ctx, dockerClient, ...)
```

This ensures multi-network containers work correctly on Docker API < 1.44 (Docker 24.x). Without this wrapper, redeploy would fail after destroying the old container, leaving the environment broken.

Backend build verified ✅

@greptileai please re-review

**Comment by mkaltner** [comment]
@greptileai please re-review

File: frontend/src/lib/services/container-service.ts [hunk 3]
```diff
@@ -17,6 +17,14 @@ export type ContainerListRequestOptions = SearchPaginationSortRequest & {
 	groupByProject?: boolean;
 };
 
+export interface ContainerRedeployResponse {
```
Comment (kmendell) [inline frontend/src/lib/services/container-service.ts]: I think we should just return the container like on creation and what not. Vs making a full on reponse type as we already hanlde this.

**Review by kmendell — CHANGES_REQUESTED** [review]

**Comment by mkaltner** [comment]
✅ **Addressed kmendell's feedback** in commit edf1cac3

Now returns full container object (like creation) instead of custom response type:

- Removed `ContainerRedeployResponse` / `ContainerRedeployData` types
- Returns `ContainerDetailsResponse` (same as `GetContainer`)
- Handles edge case: if fetch fails after successful redeploy, returns minimal response with just ID so frontend can still navigate
- Frontend updated to use `data.id` instead of `data.data.containerId`

@greptileai please re-review
